firedrake/slate/slac/compiler.py

```diff
@@ -530,6 +599,324 @@ def parenthesize(arg, prec=None, parent=None):
+
+            return (self.copy(name_in_target=name_in_target,
+                              arg_id_to_dtype={-1: NumpyType(mat_dtype), 0: NumpyType(mat_dtype), 1: NumpyType(int)}),
+                    callables_table)
+
+        def emit_call_insn(self, insn, target, expression_to_code_mapper):
+            assert self.is_ready_for_codegen()
+
+            assert isinstance(insn, loopy.CallInstruction)
+
+            parameters = insn.expression.parameters
+
+            parameters = list(parameters)
+            par_dtypes = [self.arg_id_to_dtype[i] for i, _ in enumerate(parameters)]
+
+            parameters.append(insn.assignees[0])
+            par_dtypes.append(self.arg_id_to_dtype[-1])
+
+            mat_descr = self.arg_id_to_descr[0]
+
+            arg_c_parameters = [
+                expression_to_code_mapper(
+                    par,
+                    PREC_NONE,
+                    dtype_to_type_context(target, par_dtype),
+                    par_dtype
+                ).expr
+                if isinstance(par, SubArrayRef) else
+                expression_to_code_mapper(
+                    par,
+                    PREC_NONE,
+                    dtype_to_type_context(target, par_dtype),
+                    par_dtype
+                ).expr
+                for par, par_dtype in zip(parameters, par_dtypes)
+            ]
+            c_parameters = []
+            c_parameters.insert(0, arg_c_parameters[-1])  # t1
+            c_parameters.insert(1, arg_c_parameters[0])  # t0
+            c_parameters.insert(2, mat_descr.shape[0])  # n
+            return var(self.name_in_target)(*c_parameters), False
+
+        def generate_preambles(self, target):
+            assert isinstance(target, CTarget)
+            inverse_preamble = """
+                #include <string.h>
+                #include <stdio.h>
+                #include <stdlib.h>
+                #ifndef Inverse_HPP
+                #define Inverse_HPP
+                void inverse_(PetscScalar* Aout, PetscScalar* A, PetscBLASInt N)
+                {
+                    PetscBLASInt info;
+                    PetscBLASInt* ipiv=(PetscBLASInt*) malloc(N*sizeof(PetscBLASInt));
+                    PetscScalar* Awork=(PetscScalar*) malloc(N*N*sizeof(PetscScalar));
+                    memcpy(Aout,A,N*N*sizeof(PetscScalar));
+                    LAPACKgetrf_(&N,&N,Aout,&N,ipiv,&info);
+                    if(info==0)
+                        LAPACKgetri_(&N,Aout,&N,ipiv,Awork,&N,&info);
+                    if(info!=0)
+                        fprintf(stderr,\"Getri throws nonzero info.\");
+                }
+                #endif
+            """
+            yield("lapack_inverse", "#include <petscsystypes.h>\n#include <petscblaslapack.h>\n"+inverse_preamble)
+            return
+
+    def inv_fn_lookup(target, identifier):
+        if identifier == 'inv':
+            return INVCallable(name='inv')
+
+        return None
```

Comment:
Maybe I do not understand you. What do you mean with every time around? The lookup function is defined for each slate kernel (=once per compile_expression), I do not think we can avoid that. The signature for the loopy registering is the following, I think the only thing I could potentially make nicer here is to put it in a lambda function.
```
 def register_function_id_to_in_knl_callable_mapper(program,
        func_id_to_in_knl_callable_mapper):
    """
    Returns a copy of *program* with the *function_lookup* registered.

    :arg func_id_to_in_knl_callable_mapper: A function of signature ``(target,
        identifier)`` returning a
        :class:`loopy.kernel.function_interface.InKernelCallable` or *None* if
        the *function_identifier* is not known.
    """
```

Reply:
If you want to avoid to redefine this function for every slate kernel, we would need to move it into PyOP2. In that case we would not be able to generate C-code before giving the kernel back to PyOP2, I think.
Under the assumption that we would be able to do proper inlining PyOP2 (=not transform the loopy kernel to c before we give it back to PyOP2), there should also be no C-code repetition for the C inverse function.

Reply:
Maybe we should move those Callables and lookup functions into rep2loopy in PyOP2 anyways, see https://github.com/OP2/PyOP2/blob/8abfd9556504037b0c7d5631f6c718473cb99ba0/pyop2/codegen/rep2loopy.py#L46. There is also e.g. the PetscCallable defined.

Reply:
> Maybe I do not understand you. What do you mean with every time around? 

So it doesn't look like the definition of `inv_fn_lookup` or  `INVCallable` depend on `loopy_merged` at all. So my question was why not do:
```python
class INVCallable(...):
   ...
def inv_fn_lookup(...):
   ...
```
At the top-level and then (once per compilation of the loopy kernel, just do:

```
merged = loopy.register_function_id_to...(merged, inv_fn_lookup)
```

Or am I missing something?

Reply:
Okay yes I agree, moved it to PyOP2 in https://github.com/OP2/PyOP2/pull/581